src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentDesignTimeNodeWriter.cs

```diff
@@ -596,6 +622,90 @@ public override void WriteComponentAttribute(CodeRenderingContext context, Compo
         context.CodeWriter.WriteLine();
     }
 
+    private void WritePropertyAccess(CodeRenderingContext context, ComponentAttributeIntermediateNode node, ComponentIntermediateNode componentNode, string typeInferenceLocalName)
+    {
+        if (node?.TagHelper?.Name is null || node.Annotations["OriginalAttributeSpan"] is null)
+        {
+            return;
+        }
+
+        // Write the name of the property, for rename support. 
+        // __o = nameof(global::ComponentName.PropertyName);
+        var originalAttributeName = node.Annotations[ComponentMetadata.Common.OriginalAttributeName]?.ToString() ?? node.AttributeName;
+
+        int offset;
+        if (originalAttributeName == node.PropertyName)
+        {
+            offset = 0;
+        }
+        else if (originalAttributeName.StartsWith($"@bind-{node.PropertyName}", StringComparison.Ordinal))
+        {
+            offset = 5;
+        }
+        else
+        {
+            return;
+        }
+
+        var attributeSourceSpan = (SourceSpan)node.Annotations["OriginalAttributeSpan"];
+        attributeSourceSpan = new SourceSpan(attributeSourceSpan.FilePath, attributeSourceSpan.AbsoluteIndex + offset, attributeSourceSpan.LineIndex, attributeSourceSpan.CharacterIndex + offset, node.PropertyName.Length, attributeSourceSpan.LineCount, attributeSourceSpan.CharacterIndex + offset + node.PropertyName.Length);
+
+        context.CodeWriter.Write(DesignTimeVariable);
+        context.CodeWriter.Write(" = ");
+        context.CodeWriter.Write("nameof(");
+
+        if (componentNode.TypeInferenceNode == null)
+        {
+            context.CodeWriter.Write("global::");
+            context.CodeWriter.Write(componentNode.Component.GetTypeNamespace());
+            context.CodeWriter.Write(".");
+            context.CodeWriter.Write(componentNode.Component.GetTypeNameIdentifier());
+            if (componentNode.Component.IsGenericTypedComponent())
+            {
+                context.CodeWriter.Write("<");
+                var typeArgumentCount = componentNode.Component.GetTypeParameters().Count();
+                for (var i = 0; i < typeArgumentCount; i++)
+                {
+                    if (i > 0)
+                    {
+                        context.CodeWriter.Write(",");
+                    }
+                    // We need to specify a type for the generic type parameters, but it doesn't matter which one, as we're just
+                    // getting the property name from nameof. If nameof ever supports open generics (ie, nameof(List<>.Count) then we
+                    // can remove this.
+                    context.CodeWriter.Write("string");
```

Comment:
And if the generic type has constraints `string` can't satisfy?

Reply:
Updated. In the existing tests that specify type contraints, type inference is used so this doesn't matter, but I've added a test with type constraints that doesn't use type inference, and updated the code to work in that situation.